lib/hal/impression/rpihal.ex

```diff
@@ -0,0 +1,319 @@
+      0 -> :halt
+    end
+  end
+
+  defp log(msg) when is_binary(msg) do
+    IO.puts(msg)
+    Logger.info(msg)
+  end
+
+  defp log(_state, msg) when is_binary(msg) do
+    IO.puts(msg)
+    Logger.info(msg)
+    state
+  end
+
+  defp do_update(state, display, border, buffer) do
+    IO.inspect(buffer, label: "buffer (rpihal.ex:138)")
+    Logger.info("border: #{inspect(border)}")
+    border = :red
+
+    state
+    |> log("setting resolution")
+    |> set_resolution(display.packed_resolution)
+    |> log("setting panel")
+    |> set_panel()
+    |> log("setting power")
+    |> set_power()
+    |> log("setting pll")
+    |> set_pll_clock_frequency()
+    |> log("set tse register")
+    |> set_tse_register()
+    |> log("set vcom")
+    |> set_vcom_data_interval_setting(border)
+    |> log("set gate")
+    |> set_gate_source_non_overlap_period()
+    |> log("disable external flash")
+    |> disable_external_flash()
+    |> log("set pws")
+    |> set_pws_whatever_that_means()
+    |> log("power off seq")
+    |> power_off_sequence()
+    |> log("push pixels")
+    |> push_pixel_buffer(buffer)
+    |> log("await")
+    |> await_device()
+    |> log("pon")
+    |> pon()
+    |> log("await")
+    |> await_device()
+    |> log("drf")
+    |> drf()
+    |> log("await")
+    |> await_device()
+    |> log("pof")
+    |> pof()
+    |> log("await")
+    |> await_device()
+    |> log("done")
+
+    {:ok, %State{state | setup?: true}}
+  end
+
+  #
+  # "routines" and serial commands
+  #
+
+  defp reset(state) do
+    state
+    |> set_reset(0)
+    |> sleep(100)
+    |> set_reset(1)
+    |> sleep(100)
+  end
+
+  defp soft_reset(state), do: write_command(state, 0x12)
+
+  # >HH struct.pack, so big-endian, unsigned-sort * 2
+  defp set_resolution(state, packed_resolution),
+    do: write_command(state, @tres, packed_resolution)
+
+  # Panel Setting
+  # 0b11000000 = Resolution select, 0b00 = 640x480, our panel is 0b11 = 600x448
+  # 0b00100000 = LUT selection, 0 = ext flash, 1 = registers, we use ext flash
+  # 0b00010000 = Ignore
+  # 0b00001000 = Gate scan direction, 0 = down, 1 = up (default)
+  # 0b00000100 = Source shift direction, 0 = left, 1 = right (default)
+  # 0b00000010 = DC-DC converter, 0 = off, 1 = on
+  # 0b00000001 = Soft reset, 0 = Reset, 1 = Normal (Default)
+  defp set_panel(state), do: write_command(state, @psr, [0b11101111, 0x08])
+
+  defp set_power(state),
+    do:
+      write_command(state, @pwr, [
+        Bitwise.bor(
+          Bitwise.bor(
+            Bitwise.bor(
+              # ??? - not documented in UC8159 datasheet
+              Bitwise.<<<(0x06, 3),
+              # SOURCE_INTERNAL_DC_DC
+              Bitwise.<<<(0x01, 2)
+            ),
+            # GATE_INTERNAL_DC_DC
+            Bitwise.<<<(0x01, 1)
+          ),
+          # LV_SOURCE_INTERNAL_DC_DC
+          0x01
+        ),
+        # VGx_20V
+        0x00,
+        # UC8159_7C
+        0x23,
+        # UC8159_7C
+        0x23
+      ])
```

Comment:
Hahaha, omg, what is this :D. This should probably be cleaned up with parens, `use Bitwise` and the `|||` operator.

Reply:
Wait a minute... if the shifting/bor hackery is done to build a complex binary by glueing together values, we can do that in a much simpler binary expression.